services/CurrentTime/include/ble-service-current-time/CurrentTimeService.h

```diff
@@ -0,0 +1,178 @@
+#define BLE_CURRENT_TIME_SERVICE_H
+
+#ifdef BLE_FEATURE_GATT_SERVER
+
+#include "ble/BLE.h"
+#include "ble/Gap.h"
+#include "events/EventQueue.h"
+#include "mbed_rtc_time.h"
+
+#include <time.h>
+
+/**
+ * BLE Current Time Service
+ *
+ * @par purpose
+ *
+ * @par usage
+ *
+ * @note The specification for the current time service can be found here:
+ * https://www.bluetooth.com/specifications/gatt
+ *
+ * @attention The user should not instantiate more than a single current time service service
+ */
+class CurrentTimeService : private ble::Gap::EventHandler {
+public:
+    struct CurrentTime {
+        uint16_t year;
+        uint8_t  month;
+        uint8_t  day;
+        uint8_t  hours;
+        uint8_t  minutes;
+        uint8_t  seconds;
+
+        uint8_t  weekday;
+
+        uint8_t  fractions256;
+
+        uint8_t  adjustReason;
+    };
+
+    struct EventHandler {
+        /**
+         * On current time changed
+         *
+         * This function is called if the current time characteristic is changed by the client
+         *
+         * @attention This is an abstract function and should be overridden by the user.
+         */
+        virtual void on_current_time_written(time_t current_time) {}
+    };
+
+    /**
+     * Set event handler
+     *
+     * @param handler EventHandler object to handle events raised by the current time service
+     */
+    void set_event_handler(EventHandler *handler) {
+
+    }
+
+    /**
+     * Construct and initialize a current time service.
+     *
+     * @param ble
+     * @param event_queue
+     * @param chainable_gap_event_handler
+     */
+    CurrentTimeService(BLE &ble, events::EventQueue &event_queue) :
+            _ble(ble),
+            _event_queue(event_queue),
+            _current_time_char(GattCharacteristic::UUID_CURRENT_TIME_CHAR, &_current_time) {
+        memset(&_current_time, 0, sizeof(_current_time));
+    }
+
+    ble_error_t init() {
+        GattCharacteristic *charTable[] = {&_current_time_char};
+        GattService currentTimeService(GattService::UUID_CURRENT_TIME_SERVICE, charTable, 1);
+
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+
+        ble_error_t bleError = _ble.gattServer().addService(currentTimeService);
+
+        return bleError;
+    }
+
+    time_t get_current_time() {
+        time_t t = time(NULL);
+
+        return _current_time_offset + t;
+    }
+
+    void set_time(time_t new_time) {
+        time_t t = time(NULL);
+
+        _current_time_offset = t - new_time;
+    }
+
+private:
+    void set_current_time() {
+        timeval tv{};
+        gettimeofday(&tv, NULL);
+        // printf("%u\r\n", (unsigned int) tv.tv_sec);
+
+        struct tm *current_time;
+        time_t local_time = tv.tv_sec + _current_time_offset;
+        current_time = localtime(&local_time);
```

Comment:
the variable names are all a bit close, how about something more verbose
```suggestion
        new_time_of_day = localtime(&new_time_in_seconds);
```